### ISSUE TITLE
GCB: upload "latest" build archive for every build

### ISSUE THREAD

**Issue by Dor1s** [google/oss-fuzz#3378]
This would be very useful for CIFuzz regression testing mechanism. Might have some other applications too.

**Comment by Dor1s** [comment]
@Leo-Neat 

see https://github.com/google/oss-fuzz/blob/9093fee1cb917221e3225ff9d2d9be969a7a3a8f/infra/gcb/build_project.py#L160

- use `'latest'` instead of `ts` (put it in a new variable)
- compute `upload_url_latest` similarly to https://github.com/google/oss-fuzz/blob/9093fee1cb917221e3225ff9d2d9be969a7a3a8f/infra/gcb/build_project.py#L166
- duplicate the upload binaries step with the latest url https://github.com/google/oss-fuzz/blob/master/infra/gcb/build_project.py#L310

I can help to test it then, but let's check if @oliverchang is supportive of this approach

**Comment by Dor1s** [comment]
//cc @jonathanmetzman too, in case he has a better idea or just see any problems with this one

**Comment by oliverchang** [comment]
Is listing the bucket to find the latest build (as we do on CF) not sufficient? We should avoid duplicating builds, which can be quite large. If we really do need something like this we can upload a file pointing to the latest instead.

**Comment by Dor1s** [comment]
I think listing is slow and unnecessarily complicates the code (at least in CIFuzz case).

Having a well-known URL is super convenient because it can be downloaded via HTTPS.

> We should avoid duplicating builds, which can be quite large.

There are hundreds or even thousands of builds for each project-engine-sanitizer config. I'm proposing to add just one more (which can be limited to libfuzzer-asan by the way), which will be overwritten with every newer build. I don't think it would incur much of an overhead or any significant cost otherwise.

**Comment by Dor1s** [comment]
we do this for corpus backups via `latest.zip` file

**Comment by jonathanmetzman** [comment]
> I think listing is slow and unnecessarily complicates the code (at least in CIFuzz case).
> 
> Having a well-known URL is super convenient because it can be downloaded via HTTPS.
> 
> > We should avoid duplicating builds, which can be quite large.
> 
> There are hundreds or even thousands of builds for each project-engine-sanitizer config. I'm proposing to add just one more (which can be limited to libfuzzer-asan by the way), which will be overwritten with every newer build. I don't think it would incur much of an overhead or any significant cost otherwise.

I think this is a very strong point. But, we will have to do this in other cases as well (corpus archive, coverage (unless we do it already)?, perhaps in the regression archive/bucket I'm planning on implementing in CF for use by CIF)

**Comment by oliverchang** [comment]
> I think listing is slow and unnecessarily complicates the code (at least in CIFuzz case).
> 
> Having a well-known URL is super convenient because it can be downloaded via HTTPS.
> 
> > We should avoid duplicating builds, which can be quite large.
> 
> There are hundreds or even thousands of builds for each project-engine-sanitizer config. I'm proposing to add just one more (which can be limited to libfuzzer-asan by the way), which will be overwritten with every newer build. I don't think it would incur much of an overhead or any significant cost otherwise.

I see this as an argument we should avoid adding more :) Doing this via e.g. a metadata file is equally easy right? 

And yes we do this for corpus backups, but they're significantly smaller than some of our larger builds.

**Comment by oliverchang** [comment]
Actually, I just remembered this: https://cloud.google.com/storage/docs/gsutil/commands/compose. That might work for "free". 

There may still be unintended side effects on CF though, in case we have a build pattern which incorrectly matches this 'latest' zip.

**Comment by Dor1s** [comment]
Having a metadata file pointing to the latest build SGTM!

Using any GCS API that requires extra dependencies and authentication is too much effort IMO.

**Comment by Dor1s** [comment]
> I see this as an argument we should avoid adding more :) Doing this via e.g. a metadata file is equally easy right?

Not sure. Having a file with a constant filename is a simple and reliable solution, from both the provider (OSS-Fuzz) and consumer (CIFuzz) persperctive. We just upload and download a file. No more assumptions or logic to keep in sync.

With metadata file we have to agree whether we put a `gs://`, or `https://` or GCS path (relative to the bucket). If we do the latter, we should make sure not to mess up with the leading `/` in the path. This logic must be followed in two places (OSS-Fuzz and CIFuzz) and be synchronized. If we ever change it, we'll have to make changes in two places simultaneously. And there still be a chance that we eventually change something on OSS-Fuzz end and forget to change in CIFuzz, as we don't have any tests for the GCB scripts.

That said, I'm still fine with the metadata file approach, if we think that additional 0.02 - 0.1% of the storage used is concerning.

**Comment by Dor1s** [comment]
> Actually, I just remembered this: https://cloud.google.com/storage/docs/gsutil/commands/compose. That might work for "free".

Looks like it's not different from `gsutil cp` for our purpose, as it would just copy the object:

```
$ gsutil du -hs gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/
163.7 MiB    gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer

$ gsutil compose gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/latest.zip gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/test_compose.zip
Composing gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/test_compose.zip from 1 component object(s).

$ gsutil du -hs gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/
165.33 MiB   gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer

$ gsutil du -hs gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/latest.zip
1.63 MiB     gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/latest.zip\

$ gsutil du -hs gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/test_compose.zip
1.63 MiB     gs://yara-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/yara_dex_fuzzer/test_compose.zip
```

**Comment by Dor1s** [comment]
> There may still be unintended side effects on CF though, in case we have a build pattern which incorrectly matches this 'latest' zip. That may cause exceptions.

Good call, but we should be good as per https://github.com/google/clusterfuzz/blob/7468b0de57c962694a1029655bd662ec59583b4f/src/appengine/handlers/cron/project_setup.py#L50

**Comment by oliverchang** [comment]
If we introduce such a convention, it should be generalised such that CF can use it as well in the future, rather than something specific to CIFuzz. 

CF needs to know the current revision -- which a metadata file will provide us. If we go with -latest.zip, CF will no have idea what revision it is.

Also re composite objects it looks like they're billed as separate objects so they're not "free" as I thought: https://cloud.google.com/storage/docs/composite-objects

**Comment by oliverchang** [comment]
> With metadata file we have to agree whether we put a gs://, or https:// or GCS path (relative to the bucket). If we do the latter, we should make sure not to mess up with the leading / in the path. This logic must be followed in two places (OSS-Fuzz and CIFuzz) and be synchronized. If we ever change it, we'll have to make changes in two places simultaneously. And there still be a chance that we eventually change something on OSS-Fuzz end and forget to change in CIFuzz, as we don't have any tests for the GCB scripts.

I would propose just doing a `latest.version` file or something similar in the same "directory" as the build zips, which just specify the filename (no other dir components).

**Comment by Dor1s** [comment]
> coverage (unless we do it already)?

yeah, coverage works in a similar way, there is `latest_report_info` metadata file pointing to the latest coverage data available
